Instructor: edit instructors: ensure at least one instructor is visible to students #2515

bjrodrig opened the pull request:
Fixes #2515

Changed it so that if no other instructors are visible to students, you cannot make the last visible instructor invisible. 

Ready for review

chowyb commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +63,25 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+        // We need to account for the instructor we're trying to make false.
+        if (!isDisplayedToStudents) {
+            numOfInstrDisplayed--;
+        }
```
This check seems to result in a bug: if there are two instructors in a course, one visible and one non-visible, and we attempt to modify the non-visible instructor (perhaps by changing the instructor's permissions for instance), but not modifying the visibility, this will show an error to the user.

bjrodrig commented:
Hi @chowyb. I have made the requested changes and updated the tests to reflect changes made. This is ready for review again. Thanks.

chowyb commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -1291,6 +1291,8 @@ private Const() {
                 "An instructor with the same email address already exists in the course.";
         public static final String COURSE_INSTRUCTOR_EDITED = "The changes to the instructor %s has been updated.";
+        public static final String COURSE_INSTRUCTOR_NO_INSTRUCTOR_DISPLAYED = "At least one instructor must be "
+                                                                            + "displayed to students.";
```
Avoid custom indentation in code. We've changed out from using custom indentation to using 8 spaces additional indentation due to various reasons, such as having to introduce extra changes when the length of a line changes (eg. renaming a variable/method). As an example, refer to the line for `COURSE_INSTRUCTOR_EXISTS` just above. This applies to multiple areas.

Also, if using this approach, for this case it may be better to put the entire string in a line on its own, as opposed to breaking it into two lines.

chowyb commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +72,26 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+    changes to the instructor information were successfully processed. Thus, if we are
+    trying to make the instructor false, we need to account for that.
+     */
+    protected int getNumberOfInstructorsDisplayedToStudents(String courseId, boolean currentlyDisplayed,
```
For comments for a method, use the javadoc style for the asterisks, ie:

```java
/**
 * A comment about the method
 * spanning multiple lines
 */
```

This allows most IDEs supporting Java to correctly classify it as a comment for the method, allowing it to be placed in areas where it would be helpful (eg. as a tooltip on hover for Eclipse).

chowyb commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -577,6 +577,48 @@ private void testEditInstructorAction() throws Exception {
+        and you try to uncheck the box for "Display to students as:"
+        The information in "verifyInstructorDetails" method below comes from InstructorCourseEditPageUiTest.json,
+        googleId: InsCrsEdit.tutor
+         */
```
For comments spanning multiple lines within a method, use multiple single-line comments instead ie.

```java
// a comment on what this code does
// spanning multiple lines
```

This reduces the complications involved with commenting out multiple lines of code during development later on (should that case happen).

chowyb commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +72,26 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+    trying to make the instructor false, we need to account for that.
+     */
+    protected int getNumberOfInstructorsDisplayedToStudents(String courseId, boolean currentlyDisplayed,
+                                                            boolean isDisplayedToStudents) {
```
This variable may need renaming: as there are two variables involving whether the instructor is displayed to students, it needs to be more clear what this refers to (the old display state of the instructor, or the new one should the changes be processed).

chowyb commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -28,11 +27,29 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+        boolean isDisplayedToStudents = getRequestParamValue(Const.ParamsNames.INSTRUCTOR_IS_DISPLAYED_TO_STUDENT) != null;
+        int numOfInstrDisplayed = getNumberOfInstructorsDisplayedToStudents(courseId, instructorToEdit.isDisplayedToStudents,
+                                                                             isDisplayedToStudents);
+        /* This if statement should only be processed if only one instructor is displayed to students and you are
```
Try leaving a blank line between logical blocks of code (eg. here, in between getting the instructor, and checking for the number of instructors still being displayed). This helps developers to more easily read the code.

chowyb commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -675,7 +717,7 @@ private void testDeleteInstructorAction() {
         ______TS("test the only registered instructor with the privilege to modify instructors cannot be deleted");
 
-        // Create an registered instructor with all privileges except modifying instructors
+        // Create a registered instructor with all privileges except modifying instructors
```
No need to add this change to the PR; as this change is unrelated to the issue of visible instructors, adding this change may make tracing the history of this code via `git blame` (or other methods) more difficult.

bjrodrig commented:
Hi @chowyb. I made the requested updates. For changing isDisplayedToStudents, I changed "currentlyDisplayed" to "previousStateInstructorDisplayed". When I tried changing "isDisplayedToStudents" it was causing style errors in other sections of the file, so I thought it would be easier to only make changes to code I introduced.

Also, I added a line on InstructorCourseEditPageUiTest.java between verifying status and changing the instructor attributes back after my test (line 615). I received an error when running static analysis, due to that line. Should I delete the line?

chowyb commented:
> For changing isDisplayedToStudents, I changed "currentlyDisplayed" to "previousStateInstructorDisplayed". When I tried changing "isDisplayedToStudents" it was causing style errors in other sections of the file, so I thought it would be easier to only make changes to code I introduced.

Yes, the idea was to change the name only in the method you added, as opposed to the entire file. Also, I believe I missed this out in the previous round of review, but names for boolean variables should follow a certain convention as laid out in our coding standards - the name "previousStateInstructorDisplayed" does not currently follow this convention.


> Also, I added a line on InstructorCourseEditPageUiTest.java between verifying status and changing the instructor attributes back after my test (line 615). I received an error when running static analysis, due to that line. Should I delete the line?

Seems like this has been addressed already, so this isn't an issue.


Looking through the code again, it looks like there are still places where custom indentation (ie. not using 8 spaces if it is a continuation of a previous line) is being used. While the use of custom indentation has indeed been done in the past, we're steadily moving towards using a standard 8 spaces indentation instead, so newer code should follow this convention.

bjrodrig commented:
Hi @chowyb. Thanks for all the feedback. I made the indentation changes and the boolean variable change. This is ready for review again.

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -29,10 +28,33 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+        }
+
+        boolean isDisplayedToStudents = getRequestParamValue(Const.ParamsNames.INSTRUCTOR_IS_DISPLAYED_TO_STUDENT)
+                != null;
```
Like this? (reason: higher level line breaks increase readability)
```java
boolean isDisplayedToStudents = 
        getRequestParamValue(Const.ParamsNames.INSTRUCTOR_IS_DISPLAYED_TO_STUDENT) != null;
```

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -29,10 +28,33 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+
+        // This if statement should only be processed if only one instructor is displayed to students and you are
+        // making an instructor that was previously displayed to students invisible
+        if (numOfInstrDisplayed < 1 && !isDisplayedToStudents && instructorToEdit.isDisplayedToStudents) {
```
In lieu of the long comment, you can introduce an explaining variable to make the code self-explanatory.
```java
boolean isSoleVisibleInstructor = numOfInstrDisplayed < 1 
boolean isMakingInstructorInvisible = !isDisplayedToStudents && instructorToEdit.isDisplayedToStudents;
if(isSoleVisibleInstructor && isMakingInstructorInvisible) {
    ...
```
By the way, why use `< 1` instead of `== 1` ?

damithc replied:
Also, you can check `isMakingInstructorInvisible` first and avoid calling the more expensive `getNumberOfInstructorsDisplayedToStudents(...)` if the result is false.

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +77,27 @@ protected ActionResult execute() throws EntityDoesNotExistException {
     }
 
+    /**
+     * This method gets the numbers of instructors that would be made visible to students if
```
1. `This method gets` -> `Returns` 
2. Wouldn't it be simpler to count the number of _currently_ visible instructors? Otherwise the method abstractions is a bit odd and the method name is misleading.

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +77,27 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+                numOfInstrDisplayed++;
+            }
+        }
+        // We need to account for the instructor we're trying to make false if the instructor is currently visible.
```
> ... instructor we're trying to make false ...

This can be phrased better. We don't make instructors false/true.

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -577,6 +577,48 @@ private void testEditInstructorAction() throws Exception {
+        courseEditPage.editInstructor(7, "Teammates Test", "InsCrsEdit.test@gmail.tmt",
+                false, "Instructor", "Co-owner");
+        courseEditPage.editInstructor(8, "Teammates Do Not Save", "InsCrsEdit.tutor@gmail.tmt",
+                false, "Instructor", "Tutor");
```
This code block can be made less fragile. Extract into a `makeOtherInstructorsInvisible(courseId, instructorId)`. The method should iterate though all instructors and make them invisible unless it is the specified instructor.

bjrodrig commented:
Thank you for the feedback @damithc. I made the requested updates. I added the method " makeOtherInstructorsInvisible" but I didn't use courseId or instructorId. I made it similar to editInstructor based on instructor number. 

This is ready for review again.

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +75,18 @@ protected ActionResult execute() throws EntityDoesNotExistException {
     }
 
+    protected int getNumberOfInstructorsDisplayedToStudents(
+            String courseId, boolean isDisplayedToStudentsBeforeChanges, boolean isDisplayedToStudentsAfterChanges) {
```
These boolean parameters are not being used. Your IDE did not flag them?

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +75,18 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+    protected int getNumberOfInstructorsDisplayedToStudents(
+            String courseId, boolean isDisplayedToStudentsBeforeChanges, boolean isDisplayedToStudentsAfterChanges) {
+        List<InstructorAttributes> instructors = logic.getInstructorsForCourse(courseId);
+        int numOfInstrDisplayed = 0;
```
Use full words. e.g. `displayedInstructorsCount`

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -577,6 +577,34 @@ private void testEditInstructorAction() throws Exception {
         courseEditPage.selectRoleForInstructor(editInstructorIndex, "Co-owner");
+
+        ______TS("Verify that one instructor must be visible to students");
+        courseEditPage.makeOtherInstructorsInvisible(8, 8);
```
Editing 8 instructors twice is very expensive, especially when done against the live server. Can we use a course with fewer instructors?

damithc commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -116,6 +116,17 @@ public void editInstructor(int instrNum, String name, String email,
         waitForPageToLoad();
     }
 
+    public void makeOtherInstructorsInvisible(int instrNum, int numOfInstructors) {
```
Both parameters sound the same.

damithc replied:
Add a header comment.

damithc commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -116,6 +116,17 @@ public void editInstructor(int instrNum, String name, String email,
 
+    public void makeOtherInstructorsInvisible(int instrNum, int numOfInstructors) {
+        for (int inst = 1; inst <= numOfInstructors; inst++) {
+            if (inst != instrNum) {
```
Use full words for names.

damithc commented:
@bjrodrig let us know when ready for review again.

bjrodrig commented:
Thanks. I'm still working on creating a new test case with fewer instructors.

bjrodrig commented:
@damithc 

I made the updates. I have a few questions. I am no longer using the makeOtherInstructorsInvisible method since I created a course with 2 instructors and set 1 to be visible and 1 to be invisible. Should I delete the makeOtherInstructorsInvisible method?

The IDE Usage document indicates that JSON files should have tabs and indent of 4 spaces. The InstructorCourseEditPageUiTest.json had indents of 2 spaces so I left it as 2 spaces. Should I change this to 4 spaces?

Thank you

damithc commented:
> I am no longer using the makeOtherInstructorsInvisible method since I created a course with 2 instructors and set 1 to be visible and 1 to be invisible. Should I delete the makeOtherInstructorsInvisible method?

Yes.

> The IDE Usage document indicates that JSON files should have tabs and indent of 4 spaces. The InstructorCourseEditPageUiTest.json had indents of 2 spaces so I left it as 2 spaces. Should I change this to 4 spaces?

Leave it as it is. It's better to keep the file consistent.
@wkurniawan07 is this a case of outdated docs?

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +73,17 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+            if (instructor.isDisplayedToStudents) {
+                numberOfInstructorsDisplayed++;
+            }
+        }
```
This is fine but there is a better way to count matching items in a list using Java 8 streams. https://www.concretepage.com/java/jdk-8/java-8-stream-concat-count-sorted-distinct-example

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -795,4 +797,28 @@ private InstructorCourseEditPage getCourseEditPage() {
+                false, "CS2105 Instructor 2", "Co-owner");
+        courseEditPage.verifyInstructorDetails(2, "CS2105 Instructor 2", "new_email_number_2@email.tmt",
+                false, "CS2105 Instructor 2", "Co-owner");
+        courseEditPage.verifyStatus(String.format(Const.StatusMessages.COURSE_INSTRUCTOR_EDITED, "CS2105 Instructor 2"));
```
Should you also verify that a _visible_ instructor can be made invisible if there are other visible instructors.

wkurniawan07 commented:
> is this a case of outdated docs?

@damithc JSON has actually been conveniently missed in the docs. Guess I can fix it now.

bjrodrig commented:
Thank you for the feedback @damithc 
This is ready for review again

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -55,6 +73,11 @@ protected ActionResult execute() throws EntityDoesNotExistException {
         return result;
     }
 
+    protected long getNumberOfInstructorsDisplayedToStudents(String courseId) {
```
can this be private?

bjrodrig replied:
updated

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -795,4 +797,39 @@ private InstructorCourseEditPage getCourseEditPage() {
         return loginAdminToPage(courseEditPageLink, InstructorCourseEditPage.class);
     }
 
+    private void testEditInstructorDisplayedToStudents() {
```
We can have multiple _test separator_ lines, like this:

```java
    private void testEditInstructorDisplayedToStudents() {

        instructorId = testData.instructors.get("InsCrsEdit.instructor1OfCS2105").googleId;
        courseId = testData.courses.get("InsCrsEdit.CS2105").getId();
        courseEditPage = getCourseEditPage();

        ______TS("verify sole visible instructor cannot be made invisible");
        courseEditPage.editInstructor(1, "Teammates Do Not Save", "insCrsEdit.instructor1@cs2105.tmt",
                false, "CS2105 Instructor 1", "Co-owner");

       courseEditPage.verifyInstructorDetails(1, "CS2105 Instructor 1",
                "insCrsEdit.instructor1@cs2105.tmt", true, "CS2105 Instructor 1", "Co-owner");
        courseEditPage.verifyStatus(String.format(Const.StatusMessages.COURSE_INSTRUCTOR_NO_INSTRUCTOR_DISPLAYED));

        ______TS("verify an invisible instructor can be made visible");
        courseEditPage.editInstructor(2, "CS2105 Instructor 2", "new_email_number_2@email.tmt",
                false, "CS2105 Instructor 2", "Co-owner");
        courseEditPage.verifyInstructorDetails(2, "CS2105 Instructor 2", "new_email_number_2@email.tmt",
                false, "CS2105 Instructor 2", "Co-owner");
        courseEditPage.verifyStatus(String.format(Const.StatusMessages.COURSE_INSTRUCTOR_EDITED, "CS2105 Instructor 2"));

         ______TS("verify an instructor can be made invisible if there are other visible instructors");
        courseEditPage.editInstructor(2, "CS2105 Instructor 2", "insCrsEdit.instructor2@cs2105.tmt",
                true, "CS2105 Instructor 2", "Co-owner");
        // set instructor 1 to invisible and also change email (for variety)
        courseEditPage.editInstructor(1, "CS2105 Instructor 1", "new_email@email.tmt", false,
                "CS2105 Instructor 1", "Co-owner");
        courseEditPage.verifyInstructorDetails(1, "CS2105 Instructor 1", "new_email@email.tmt",
                false, "CS2105 Instructor 1", "Co-owner");
        courseEditPage.verifyStatus(String.format(Const.StatusMessages.COURSE_INSTRUCTOR_EDITED, "CS2105 Instructor 1"));
    }
```
Note how some comments have been removed/edited.

bjrodrig replied:
updated

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -795,4 +797,39 @@ private InstructorCourseEditPage getCourseEditPage() {
+        // first make sure both instructors are visible
+        courseEditPage.editInstructor(2, "CS2105 Instructor 2", "insCrsEdit.instructor2@cs2105.tmt",
+                true, "CS2105 Instructor 2", "Co-owner");
+        // now set instructor 1 to false and change email
```
Why the need to change the email?

bjrodrig replied:
I wanted to show that when you make the instructor false, all other changes made are appropriately saved and reflected in the updated browser. I chose to change email instead of name, since the alphabetical order of the display name effects the instructor index when performing testing, and I wanted to make sure the order of instructor 1 and instructor 2 would not change.

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -795,4 +797,39 @@ private InstructorCourseEditPage getCourseEditPage() {
+        courseEditPage.editInstructor(1, "CS2105 Instructor 1", "new_email@email.tmt", false,
+                "CS2105 Instructor 1", "Co-owner");
+        courseEditPage.verifyInstructorDetails(1, "CS2105 Instructor 1", "new_email@email.tmt",
+                false, "CS2105 Instructor 1", "Co-owner");
```
What's the purpose of this line?

bjrodrig replied:
I wanted to show that if you change a visible instructor to invisible and make any additional changes, those additional changes will be appropriately saved to the database and reflected in the updated browser.

damithc commented on src/main/java/teammates/ui/controller/InstructorCourseInstructorEditSaveAction.java:
```diff
@@ -29,10 +28,29 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+        boolean isDisplayedToStudentsAfterChanges =
+                getRequestParamValue(Const.ParamsNames.INSTRUCTOR_IS_DISPLAYED_TO_STUDENT) != null;
+        boolean isMakingInstructorInvisible = !isDisplayedToStudentsAfterChanges
+                && instructorToEdit.isDisplayedToStudents;
```
More intuitive to read if you swap the conditions? i.e., `instructorToEdit.isDisplayedToStudents &&  !isDisplayedToStudentsAfterChanges`

bjrodrig replied:
updated

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -55,6 +55,8 @@ public void allTests() throws Exception {
 
         testSanitization();
+
+        testEditInstructorDisplayedToStudents();
```
better name `testEditingInstructorVisibility()`

bjrodrig replied:
updated

bjrodrig commented:
@damithc 
I have addressed comments and answered the two questions above. This is ready for review.

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -795,4 +797,37 @@ private InstructorCourseEditPage getCourseEditPage() {
+
+        courseEditPage.verifyInstructorDetails(2, "CS2105 Instructor 2", "new_email_number_2@email.tmt",
+                false, "CS2105 Instructor 2", "Co-owner");
+        courseEditPage.verifyStatus(String.format(Const.StatusMessages.COURSE_INSTRUCTOR_EDITED, "CS2105 Instructor 2"));
```
See if this three lines can be extracted as a private method and reused (possibly in other places of this class as well). e.g.

```java
verifyIntstrcutorEditSuccess(
        2, "CS2105 Instructor 2", "new_email_number_2@email.tmt", false, "CS2105 Instructor 2", "Co-owner")
```

damithc commented:
@bjrodrig any updated?
@chowyb reminder to have another look as the code has changed a lot since you last reviewed.

bjrodrig commented:
@damithc Not yet. I may not have time today, but I can update tomorrow.

bjrodrig commented:
@damithc I have made the requested updates. This is ready for review again.

damithc commented on src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java:
```diff
@@ -212,6 +212,12 @@ public void verifyInstructorDetails(int instrNum, String newName, String newEmai
+            boolean newIsDisplayedToStudents, String newDisplayName, String newRole, String expectedStatus) {
+        verifyInstructorDetails(instrNum, newName, newEmail, newIsDisplayedToStudents, newDisplayName, newRole);
+        verifyStatus(expectedStatus);
+    }
```
* this method can be in the `InstructorCourseEditPageUiTest.java`

* `verifyInstructorDetailsAndStatus(...)` -> `verifyInstructorEdited(...)` (no need for the name to contain all details of the method).

* `instrNum` -> `instructorIndex` (prefer to use full words)

* No need to take the status as a parameter. Use this method only for success cases (i.e., the status message is always the same). That way, the call site can be less verbose.
```java
verifyInstructorEdited(1, "CS2105 Instructor 1", "new_email@email.tmt",
                false, "CS2105 Instructor 1", "Co-owner");
```
In fact you can consider this:
```java
private void verifyInstructorEditSuccess(...){
    courseEditPage.editInstructor(...);
    verifyInstructorDetails(...);
    verifyStatus(...);
}
```

bjrodrig commented:
@damithc 
Thanks for the feedback. updates made.

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -34,6 +34,19 @@ protected void prepareTestData() {
+    /**
+     * Verifies that changes to instructor are successfully saved to the database
+     * and that the success message is displayed in the browser.
+     **/
```
Should be `*/`?

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -34,6 +34,19 @@ protected void prepareTestData() {
+     * and that the success message is displayed in the browser.
+     **/
+    private void verifyInstructorEditSuccess(int instructorIndex, String newName, String newEmail,
+                                            boolean newIsDisplayedToStudents, String newDisplayName, String newRole) {
```
Indentation should be 8 spaces here.

bjrodrig commented:
@damithc  updated

damithc commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -34,6 +34,19 @@ protected void prepareTestData() {
+     * Verifies that changes to instructor are successfully saved to the database
+     * and that the success message is displayed in the browser.
+     */
+    private void verifyInstructorEditSuccess(int instructorIndex, String newName, String newEmail,
```
1. Given the expanded scope of the method, may be the name should be `editInstructorSuccessfully()`? The header comment need to change accordingly.
2. We usually put the utility method somewhere below where it is used, not above.

bjrodrig commented:
updated

wkurniawan07 commented on src/test/java/teammates/test/cases/browsertests/InstructorCourseEditPageUiTest.java:
```diff
@@ -795,4 +789,41 @@ private InstructorCourseEditPage getCourseEditPage() {
+                false, "CS2105 Instructor 1", "Co-owner");
+        courseEditPage.verifyInstructorDetails(1, "CS2105 Instructor 1", "insCrsEdit.instructor1@cs2105.tmt",
+                true, "CS2105 Instructor 1", "Co-owner");
+        courseEditPage.verifyStatus(String.format(Const.StatusMessages.COURSE_INSTRUCTOR_NO_INSTRUCTOR_DISPLAYED));
```
No need for `String.format` here?

wkurniawan07 commented:
Hi @bjrodrig, any updates here?

bjrodrig commented:
No, I haven't had time to update this. Did you still want the tests to be changed? I thought when @damithc  reviewed, the tests were ok?

shaharyar-shamshi commented:
@bjrodrig  Sorry Actually I thought you are no more interested so I Opened pull request for this issue

bjrodrig commented:
Sure, @shaharyar-shamshi I'm not working on it

wkurniawan07 commented:
Closing as agreed.
